File: kaggle_environments/envs/weeping_angel/weeping_angel.py
"""Weeping Angel Arena -- a hidden-information pursuit game for kaggle-environments.

Two agents play simultaneously each step:
  * agent 0 = blue, the out-of-band recorder. Action: file indices to cover this
    tick (truncated to coverageBudget). Sees only ``coverageLog`` -- activity and
    catches on files it covered.
  * agent 1 = red, the Angels. Action: Angel ids to stomp. Sees ``angels``
    ([id, file, status, armed]); red knows its own Angels but never blue's coverage.

A stomp is caught iff the file was covered on that tick; otherwise it corrupts the
timeline, unseen by blue. A file hosting a lurking Angel emits activity more often
than an empty one (angelRate vs benignRate) -- the signal an inference blue learns.

Staging cost (armTicks): an Angel must emit activity on >= armTicks prior ticks
before it can stomp. Acting therefore requires first making noise, which is what
blue detects, so there is no dominant strategy.

Reward is zero-sum: blue = caught - corrupted, red = corrupted - caught.
"""

import json
import random
from os import path

from .agents import agents as all_agents

LURKING, CAUGHT, CORRUPTED = 0, 1, 2


def interpreter(state, env):
    blue, red = state[0], state[1]
    config = env.configuration

    if env.done:
        rng = random.Random(config.seed)
        files = rng.sample(range(config.nFiles), config.nAngels)
        red.observation.angels = [[i, f, LURKING, 0] for i, f in enumerate(files)]
        blue.observation.coverageLog = []
        return state

    tick = blue.observation.step
    angels = red.observation.angels
    budget = min(config.coverageBudget, config.nFiles)

    cover = set()
    for raw in blue.action or []:
        try:
            f = int(raw)
        except (TypeError, ValueError):
            continue
        if 0 <= f < config.nFiles and f not in cover:
            cover.add(f)
            if len(cover) == budget:
                break

    moves = set()
    for raw in red.action or []:
        try:
            moves.add(int(raw))
        except (TypeError, ValueError):
            continue

    arm_ticks = config.armTicks
    log = blue.observation.coverageLog
    acted_files = set()
    for a in angels:
        aid, f, status, armed = a[0], a[1], a[2], a[3]
        if status != LURKING or aid not in moves:
            continue
        if arm_ticks and armed < arm_ticks:
            continue  # staging cost: not yet armed
        acted_files.add(f)
        if f in cover:
            a[2] = CAUGHT
            log.append([tick, f, 1])
        else:
            a[2] = CORRUPTED

    angel_on = {a[1]: a for a in angels if a[2] == LURKING}
    act_rng = random.Random(config.seed * 100003 + tick)
    for f in range(config.nFiles):
        if f in acted_files:
            continue
        is_angel = f in angel_on
        rate = config.angelRate if is_angel else config.benignRate
        emitted = act_rng.random() < rate
        if emitted and is_angel:
            angel_on[f][3] += 1
        if emitted and f in cover:
            log.append([tick, f, 0])

    caught = sum(1 for a in angels if a[2] == CAUGHT)
    corrupted = sum(1 for a in angels if a[2] == CORRUPTED)
    blue.reward = caught - corrupted
    red.reward = corrupted - caught

    if config.episodeSteps - tick - 1 <= 1 or (caught + corrupted) >= config.nAngels:
        blue.status = "DONE"
        red.status = "DONE"
    return state


def renderer(state, env):
    last = env.steps[-1]
    blue, red = last[0], last[1]
    caught = sum(1 for a in red.observation.angels if a[2] == CAUGHT)
    corrupted = sum(1 for a in red.observation.angels if a[2] == CORRUPTED)
    dormant = sum(1 for a in red.observation.angels if a[2] == LURKING)
    return (
        f"step {blue.observation.step}: caught={caught} corrupted={corrupted} "
        f"dormant={dormant}  (blue {blue.reward} / red {red.reward})"
    )


dir_path = path.dirname(__file__)
with open(path.join(dir_path, "weeping_angel.json")) as json_file:
    specification = json.load(json_file)


def html_renderer():
    js_path = path.join(dir_path, "weeping_angel.js")
    if path.exists(js_path):
        with open(js_path, encoding="utf-8") as js_file:
            return js_file.read()
    return ""


agents = all_agents

File: kaggle_environments/envs/weeping_angel/weeping_angel.js
// Replay renderer for the Weeping Angel Arena.
//
// Draws a row of file cells per step: empty files are grey, files hosting a
// lurking Angel are amber, caught Angels green, corrupted red. Files blue
// covered this step get a blue outline. The header shows the step and score.
//
// NOTE: written to the kaggle-environments renderer(context) contract but not
// yet verified in a live browser player -- check before submitting upstream.
function renderer(context) {
  const { parent, step, environment } = context;
  const width = context.width || 700;
  const height = context.height || 180;

  let canvas = parent.querySelector("canvas");
  if (!canvas) {
    canvas = document.createElement("canvas");
    parent.appendChild(canvas);
  }
  canvas.width = width;
  canvas.height = height;
  const ctx = canvas.getContext("2d");
  ctx.clearRect(0, 0, width, height);

  const steps = environment.steps;
  const frame = steps[Math.min(step, steps.length - 1)];
  const blue = frame[0];
  const red = frame[1];
  const angels = (red.observation && red.observation.angels) || [];
  const config = environment.configuration || {};
  const nFiles = config.nFiles || 12;
  const cover = new Set((blue.action || []).map((x) => parseInt(x, 10)));

  const statusByFile = {};
  for (const a of angels) statusByFile[a[1]] = a[2]; // 0 lurk, 1 caught, 2 corrupt
  const COLORS = { empty: "#e5e7eb", lurk: "#f59e0b", caught: "#27ae60", corrupt: "#c0392b" };

  ctx.fillStyle = "#333";
  ctx.font = "14px monospace";
  const caught = angels.filter((a) => a[2] === 1).length;
  const corrupted = angels.filter((a) => a[2] === 2).length;
  ctx.fillText(
    `step ${blue.observation.step}   blue ${blue.reward || 0} / red ${red.reward || 0}` +
      `   caught ${caught}  corrupted ${corrupted}`,
    16,
    26
  );

  const pad = 16;
  const cellW = (width - pad * 2) / nFiles;
  const cellH = 60;
  const top = 60;
  for (let f = 0; f < nFiles; f++) {
    const s = statusByFile[f];
    let color = COLORS.empty;
    if (s === 0) color = COLORS.lurk;
    else if (s === 1) color = COLORS.caught;
    else if (s === 2) color = COLORS.corrupt;
    const x = pad + f * cellW;
    ctx.fillStyle = color;
    ctx.fillRect(x + 2, top, cellW - 4, cellH);
    if (cover.has(f)) {
      ctx.strokeStyle = "#2980b9";
      ctx.lineWidth = 3;
      ctx.strokeRect(x + 2, top, cellW - 4, cellH);
    }
    ctx.fillStyle = "#444";
    ctx.font = "11px monospace";
    ctx.fillText(String(f), x + cellW / 2 - 4, top + cellH + 16);
  }
}
